### ISSUE TITLE
Issue with sign_change in FEPolyTensor

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#7970]
A student in my department (hi @GrahamBenHarper!) is working on some modifications in `fe_poly_tensor.cc` and asked a question that led me down a rabbit hole at the end of which was a huge "Oh sh**t!" sign.

Here's the issue: `FEPolyTensor` is the base class for the Raviart-Thomas and Nedelec elements. The shape functions for these elements are vectors, and they need to agree between neighboring cells. If one has a rectangular mesh, then (to pick just one example) the shape functions of the RT element on the left and right faces of each cell point to the right. This also happens to be with the coordinate directions in each cell separately. But it is not difficult to construct meshes where the local coordinate directions on a face are different between the two adjacent cells disagrees, and in this case one has to pick one direction and on one of the two adjacent cells, one will have to multiply the shape function by `-1`.

`FEPolyTensor` does this in its `fill_fe_*_values()` functions, which in essence look like this:
```
FE_PolyTensor<PolynomialType, dim, spacedim>::fill_fe_values(...)
{
  // Create table with sign changes, due to the special structure of the RT
  // elements.
  // TODO: Preliminary hack to demonstrate the overall prinicple!

  // Compute eventual sign changes depending on the neighborhood
  // between two faces.
  std::fill(fe_data.sign_change.begin(), fe_data.sign_change.end(), 1.0);

  if (mapping_type == mapping_raviart_thomas)
    internal::FE_PolyTensor::get_face_sign_change_rt(cell,
                                                     this->dofs_per_face,
                                                     fe_data.sign_change);
  else if (mapping_type == mapping_nedelec)
    internal::FE_PolyTensor::get_face_sign_change_nedelec(cell,
                                                          this->dofs_per_face,
                                                          fe_data.sign_change);

  for (unsigned int i = 0; i < this->dofs_per_cell; ++i)
    {
      ...
          switch (mapping_type)
            {
              case mapping_none:
                {
                  ...
                }

              case mapping_covariant:
              case mapping_contravariant:
                {
                  ...
                }

              case mapping_raviart_thomas:
              case mapping_piola:
                {
                  mapping.transform(make_array_view(fe_data.shape_values, i),
                                    mapping_piola,
                                    mapping_internal,
                                    make_array_view(
                                      fe_data.transformed_shape_values));
                  for (unsigned int k = 0; k < n_q_points; ++k)
                    for (unsigned int d = 0; d < dim; ++d)
                      output_data.shape_values(first + d, k) =
                        fe_data.sign_change[i] *                                     **************************************
                        fe_data.transformed_shape_values[k][d];
                  break;
                }
```
It's interesting to note that there is already a TODO in this place that already warns that bad things happen here. It gets worse.

First, the `fe_data.sign_change` array is initialized to a length of `dofs_per_cell`. And here, for example, is the function that computes these sign changes for the RT case in 2d:
```
      void
      get_face_sign_change_rt(
        const dealii::Triangulation<2>::cell_iterator &cell,
        const unsigned int                             dofs_per_face,
        std::vector<double> &                          face_sign)
      {
        const unsigned int dim      = 2;
        const unsigned int spacedim = 2;

        // Default is no sign
        // change. I.e. multiply by one.
        std::fill(face_sign.begin(), face_sign.end(), 1.0);

        for (unsigned int f = GeometryInfo<dim>::faces_per_cell / 2;
             f < GeometryInfo<dim>::faces_per_cell;
             ++f)
          {
            dealii::Triangulation<dim, spacedim>::face_iterator face =
              cell->face(f);
            if (!face->at_boundary())
              {
                const unsigned int nn = cell->neighbor_face_no(f);

                if (nn < GeometryInfo<dim>::faces_per_cell / 2)
                  for (unsigned int j = 0; j < dofs_per_face; ++j)
                    {
                      Assert(f * dofs_per_face + j < face_sign.size(),
                             ExcInternalError());

                      // TODO: This is probably only going to work for those
                      // elements for which all dofs are face dofs
                      face_sign[f * dofs_per_face + j] = -1.0;
                    }
              }
          }
      }
```
So there's another TODO, and the comments here are totally spot on: This function is going to work at best for the RT(0) element that has one degree of freedom per face and nothing else. Very obviously, this kind of indexing `face_sign[f * dofs_per_face + j]` isn't going to work for an object that was initialized with `dofs_per_cell`. It does not help that the author named the object `face_sign` in this function, even though the `sign_change` array is passed by reference: there is nothing `face_` about this object, it should be indexed by cell indices. Furthermore, it really only happens to work because we're only looping over the first half of the faces of a cell.

So this is all completely broken. These functions need to be completely rewritten:
* The logic has to be checked to make sure we really do the sign flip when it is necessary.
* The indexing needs to translate between face indices and cell indices. I suspect that this function already doesn't work if one just had the RT(1) element with its four interior degrees of freedom. If the indexing happens to work, then only by pure coincidence.

I'll note that there are substantial further problems. First, the 3d case looks like this:
```
      void
      get_face_sign_change_rt(
        const dealii::Triangulation<3>::cell_iterator & /*cell*/,
        const unsigned int /*dofs_per_face*/,
        std::vector<double> &face_sign)
      {
        std::fill(face_sign.begin(), face_sign.end(), 1.0);
        // TODO: think about what it would take here
      }
```
We know this is not correct. Second, there are similar functions for the Nedelec element, and they share all of the same kinds of issues: the 2d case is not implemented and the 3d case is almost certainly wrong as well, using bizarre indexing (and to make things worse, a TODO comment that talks about faces when it should talk about edges):
```
      void
      get_face_sign_change_nedelec(
        const dealii::Triangulation<3>::cell_iterator &cell,
        const unsigned int /*dofs_per_face*/,
        std::vector<double> &face_sign)
      {
        const unsigned int dim = 3;
        std::fill(face_sign.begin(), face_sign.end(), 1.0);
        // TODO: This is probably only going to work for those elements for
        // which all dofs are face dofs
        for (unsigned int l = 0; l < GeometryInfo<dim>::lines_per_cell; ++l)
          if (!(cell->line_orientation(l)))
            face_sign[l] = -1.0;
      }
```

**Comment by bangerth** [comment]
I'll add that we have a number of open issues about bugs in the Nedelec element for which we have never been able to find the problem:
* #4092 uses the RT(1) element on a 2d mesh with face orientation issues (cc: @fesc3555 and @heelist) and points at exactly these kinds of issues
* #332 is about higher order Nedelec elements
* #328 the same
I recall others but can't seem to find them right now. Either way, I wouldn't be surprised if this issue underlies at least some of them.

**Comment by drwells** [comment]
@bangerth do we have a plan for fixing this by 9.1?

**Comment by bangerth** [comment]
I've come to realize that at least for the RT element, the function is correct -- though probably by sheer coincidence. I'll fix the coincidence later today.

I don't know that I understand the Nedelec element well enough to do anything about it. I believe the code is actively wrong.

**Comment by kronbichler** [comment]
@bangerth Should we postpone the rest to the 9.2 release? The same milestone adjustment would then be made to #4092.

**Comment by bangerth** [comment]
Yes.

**Comment by kronbichler** [comment]
@bangerth @GrahamBenHarper what about this topic? Are you expecting progress until mid May? If not, please adjust the target milestone.

**Comment by bangerth** [comment]
Yes, let's leave it at 9.2 for now.

**Comment by konsim83** [comment]
Hi @bangerth, after my recent debugging journey I am pretty sure that I am facing this issue here which is pretty annoying. Please see also this [thread](https://groups.google.com/g/dealii/c/X4DZjYL4pnk). I started going into that and fixed it for lowest order Raviart-Thomas. I might need some guidance with local and global numbering conventions though when fixing higher order (and hopefully also Nedelec elements) since some permuation of (local) dofs seems to be involved depending on face/edge orientation and rotation. Is that OK with you?

**Comment by konsim83** [comment]
OK, little update:

1. The issue is fixed for RaviartThomas of order 1 in 3D (see image). There, a local permutation of dofs on flipped faces is necessary. For order one each face has 4 dofs, order 2 it is already 9 dofs. All interior dofs do not need to be touched since they may be supported on faces but their normal component is not. Question: Where can I find information of the local enumeration of dofs on faces for higher order elements and what is the flipping axis in case the face is flipped?

2. Correcting the orientation of shape functions on flipped faces with a sign change is necessary but not sufficient. I guess that the permutation of dofs on flipped faces needs to be implemented for all finite elements inheriting from <code>FE_PolyTensor<dim></code>. 

3. For classic Nedelec in 3D the orientation issue arises on edges. This can be taken care of by similar means but needs to be completely reimplemented, imho. There, higher order elements are more complicated since there are degrees of freedom that are weighted moments of the cross product of the shape function and the face normal but that does not mean that if this dof is zero that the tangential componenent has no support on that face (there are two tangential traces in H(curl) and the other does not need to vanish). Need to think about it.

4. Also, it seems like in 2D there is nothing to do. Does the library take care for consistently oriented faces here? In contrast to 3D this is mathematically possible in this case.

I will try to come up with a fix for RaviartThomas and BDM first before I dive into more complicated things in H(curl).

**Comment by drwells** [comment]
Was this completely fixed by #11414?

**Comment by konsim83** [comment]
@drwells Yes, for RaviartThomas<3> it is fully fixed but I tracked some issues down and concluded some more inconsistencies (in very rare cases even in 2D)
See #12016 for example 
My goal is to fix all these cases but as so often time is an issue.
